### PR TITLE
use proper make executable based on os/compiler

### PR DESCRIPTION
**Pull request by ericLemanissier**
on Visual studio, calls jom if available, otherwhise nmake
on Windows +gcc/clang, calls mingw32-make
on other configurations, calls make (no change)

fixes conan-io/conan#1254

### CODE REVIEW

**Comment by ericLemanissier** [comment]
I did not test the change yet

File: conans/client/configure_build_environment.py [hunk 1]
```diff
@@ -186,7 +187,17 @@ def make(self, args=""):
...
             cpu_count_option = ("-j%s" % cpu_count()) if "-j" not in str_args else ""
-            self._conanfile.run("make %s %s" % (str_args, cpu_count_option))
+            if self._conanfile.settings.get_safe("os") == "Windows":
+                if self._compiler == "Visual Studio":
+                    build_command = find_executable("jom.exe")
+                    if build_command:
+                        self._conanfile.run("%s %s %s" % (build_command, str_args, cpu_count_option))
+                    else:
+                        self._conanfile.run("nmake %s" % str_args)
+                else:
```
Comment (lasote) [inline conans/client/configure_build_environment.py]: Check gcc?

File: conans/client/configure_build_environment.py [hunk 2]
```diff
@@ -186,7 +187,17 @@ def make(self, args=""):
...
+                    self._conanfile.run("mingw32-make %s %s" % (str_args, cpu_count_option))
+            else:
```
Comment (lasote) [inline conans/client/configure_build_environment.py]: What about the case of msys2 with make in windows? It's a little tricky, why that specific order and no other?
Probably we need to add a parameter "make_command" to the `make` method to call explicitly to the indicated by the user.

Comment (ericLemanissier) [reply conans/client/configure_build_environment.py]: maybe the best it to use find_executable and use mingw32-make if present, otherwise make ?

Comment (lasote) [reply conans/client/configure_build_environment.py]: Yes, could be. And I probably would introduce the optional parameter to override the conan "choose" of the correct make, if needed

**Comment by SSE4** [comment]
I don't think such heuristics are really good, e.g.:

- jom doesn't handle all makefiles properly, e.g. it didn't work well for OpenSSL
- Visual Studio doesn't always imply nmake build (e.g. ffmpeg is still built via make in MSYS2)
- Windows + clang doesn't imply MinGW

I would prefer adding **make_program** argument instead

**Comment by ericLemanissier** [comment]
Ok for adding a make_program argument.
I propose the following if this argument is not set by the caller or empty: call make on non windows platform, and on windows call the first found (by find_executable) in this list mingw32-make, jom, nmake, make.

**Comment by lasote** [comment]
Hi, we suggested to @SSE4 use the spawn module as you are doing in this PR to implement the "which" tool we merged, but it looks like there is some problem in some machines, where that module is not present: 
https://github.com/conan-io/conan/pull/1801#issuecomment-334672800

Maybe we should use the new `tool.which` (merged to develop)

**Comment by memsharded** [comment]
I agree with @SSE4, I am not fully sure I like the heuristics. So if I don't specify in my recipes a make_program, and for some reasons my installation of mingw is incorrect, it might end picking up vcvars & nmake to build it?

I think we might need some kind of convention with env-vars or something to be able to define the make_program.

**Comment by ericLemanissier** [comment]
Sorry I won't have possibility to implement and test this feature

**Pull request closed without merging**